### ISSUE TITLE
istio-proxy on app pod reported "envoy.config.listener.v3.Listener rejected, has duplicate address"

### ISSUE THREAD

**Issue by soft-way** [istio/istio#43095]
### Bug Description

1. Create an istio primary cluster on dual-stack k8s (want to create a istio multiple clusters env)
2. Deploy a test applicaiton on this primary cluster (use sample app included in istio delivery package)
kubectl create namespace sample
kubectl label namespace sample istio-injection=enabled
kubectl apply -f samples/helloworld/helloworld.yaml -l service=helloworld -n sample
kubectl apply -f samples/helloworld/helloworld.yaml -l version=v2 -n sample

3. istio-proxy in helloworld-v2 reported below errors (192.168.2.127_9100' has duplicate address '192.168.2.127:9100,[2001:2:127:1::ab06]:9100):
2023-02-02T06:48:37.340830Z	info	JWT policy is third-party-jwt
2023-02-02T06:48:37.340834Z	info	using credential fetcher of JWT type in cluster.local trust domain
2023-02-02T06:48:37.441084Z	info	Workload SDS socket not found. Starting Istio SDS Server
2023-02-02T06:48:37.441093Z	info	Opening status port 15020
2023-02-02T06:48:37.441115Z	info	CA Endpoint istiod.istio-system.svc:15012, provider Citadel
2023-02-02T06:48:37.441148Z	info	Using CA istiod.istio-system.svc:15012 cert with certs: var/run/secrets/istio/root-cert.pem
2023-02-02T06:48:37.441258Z	info	citadelclient	Citadel client using custom root cert: var/run/secrets/istio/root-cert.pem
2023-02-02T06:48:37.451079Z	info	ads	All caches have been synced up in 117.214492ms, marking server ready
2023-02-02T06:48:37.451662Z	info	xdsproxy	Initializing with upstream address "istiod.istio-system.svc:15012" and cluster "app2"
2023-02-02T06:48:37.452937Z	info	starting Http service at 127.0.0.1:15004
2023-02-02T06:48:37.453006Z	info	sds	Starting SDS grpc server
2023-02-02T06:48:37.453799Z	info	Pilot SAN: [istiod.istio-system.svc]
2023-02-02T06:48:37.455270Z	info	Starting proxy agent
2023-02-02T06:48:37.455290Z	info	starting
2023-02-02T06:48:37.455305Z	info	Envoy command: [-c etc/istio/proxy/envoy-rev.json --drain-time-s 45 --drain-strategy immediate --parent-shutdown-time-s 60 --local-address-ip-version v4 --file-flush-interval-msec 1000 --disable-hot-restart --log-format %Y-%m-%dT%T.%fZ	%l	envoy %n	%v -l warning --component-log-level misc:error --concurrency 2]
2023-02-02T06:48:37.755043Z	info	cache	generated new workload certificate	latency=302.039964ms ttl=23h59m59.244965151s
2023-02-02T06:48:37.755070Z	info	cache	Root cert has changed, start rotating root cert
2023-02-02T06:48:37.755085Z	info	ads	XDS: Incremental Pushing:0 ConnectedEndpoints:0 Version:
2023-02-02T06:48:37.755128Z	info	cache	returned workload trust anchor from cache	ttl=23h59m59.244875283s
2023-02-02T06:48:37.828272Z	info	xdsproxy	connected to upstream XDS server: istiod.istio-system.svc:15012
2023-02-02T06:48:37.848718Z	info	ads	ADS: new connection for node:helloworld-v2-79bf565586-f44h7.sample-1
2023-02-02T06:48:37.848768Z	info	cache	returned workload certificate from cache	ttl=23h59m59.151236546s
2023-02-02T06:48:37.848791Z	info	ads	ADS: new connection for node:helloworld-v2-79bf565586-f44h7.sample-2
2023-02-02T06:48:37.848882Z	info	cache	returned workload trust anchor from cache	ttl=23h59m59.1511224s
2023-02-02T06:48:37.848964Z	info	ads	SDS: PUSH request for node:helloworld-v2-79bf565586-f44h7.sample resources:1 size:4.0kB resource:default
2023-02-02T06:48:37.849096Z	info	ads	SDS: PUSH request for node:helloworld-v2-79bf565586-f44h7.sample resources:1 size:1.1kB resource:ROOTCA
2023-02-02T06:48:37.875942Z	warning	envoy config	error adding listener: '192.168.2.127_9100' has duplicate address '192.168.2.127:9100,[2001:2:127:1::ab06]:9100' as existing listener
2023-02-02T06:48:37.910821Z	warning	envoy config	gRPC config for **type.googleapis.com/envoy.config.listener.v3.Listener rejected: Error adding/updating listener(s) 192.168.2.127_9100: error adding listener: '192.168.2.127_9100' has duplicate address '192.168.2.127:9100,[2001:2:127:1::ab06]:9100' as existing listener**

2023-02-02T06:48:38.033260Z	warn	Envoy proxy is NOT ready: config received from XDS server, but was rejected: cds updates: 1 successful, 0 rejected; lds updates: 0 successful, 1 rejected
2023-02-02T06:48:39.042307Z	warn	Envoy proxy is NOT ready: config received from XDS server, but was rejected: cds updates: 1 successful, 0 rejected; lds updates: 0 successful, 1 rejected
2023-02-02T06:48:39.052383Z	warn	Envoy proxy is NOT ready: config received from XDS server, but was rejected: cds updates: 1 successful, 0 rejected; lds updates: 0 successful, 1 rejected
2023-02-02T06:48:41.052128Z	warn	Envoy proxy is NOT ready: config received from XDS server, but was rejected: cds updates: 1 successful, 0 rejected; lds updates: 0 successful, 1 rejected
2023-02-02T06:48:43.052502Z	warn	Envoy proxy is NOT ready: config received from XDS server, but was rejected: cds updates: 1 successful, 0 rejected; lds updates: 0 successful, 1 rejected
2023-02-02T06:48:45.052379Z	warn	Envoy proxy is NOT ready: config received from XDS server, but was rejected: cds updates: 1 successful, 0 rejected; lds updates: 0 successful, 1 rejected
 
4.  yes, I deployed one prometheus exporter service, it used port 9100. service dump as below:
apiVersion: v1
kind: Service
metadata:
  annotations:
    prometheus.io/scrape: "true"
  creationTimestamp: "2023-02-02T06:18:37Z"
  labels:
    kubernetes.io/name: NodeExporter
  name: node-exporter
  namespace: prometheus-node-exporter
  resourceVersion: "822"
  uid: 5fe250ab-6896-4416-bf1f-796251d2aad5
spec:
  clusterIP: None
  clusterIPs:
  - None
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - name: metrics
    port: 9100
    protocol: TCP
    targetPort: 9100
  selector:
    app: prometheus
    module: node-exporter
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}

5. istiod also reported below error:
2023-02-02T06:48:26.853336Z	info	ads	RDS: PUSH for node:helloworld-v1-77cb56d4b4-l9lzr.sample resources:18 size:10.1kB cached:16/18
2023-02-02T06:48:35.024844Z	info	Sidecar injection request for sample/helloworld-v2-79bf565586-***** (actual name not yet known)
2023-02-02T06:48:35.058683Z	info	ads	Incremental push, service helloworld.sample.svc.cluster.local at shard Kubernetes/app2 has no endpoints
2023-02-02T06:48:35.159740Z	info	ads	Push debounce stable[19] 1 for config ServiceEntry/sample/helloworld.sample.svc.cluster.local: 101.008667ms since last change, 101.008634ms since last push, full=false
2023-02-02T06:48:35.159775Z	info	ads	XDS: Incremental Pushing:2023-02-02T06:48:26Z/9 ConnectedEndpoints:3 Version:2023-02-02T06:48:26Z/9
2023-02-02T06:48:37.035419Z	info	ads	Incremental push, service helloworld.sample.svc.cluster.local at shard Kubernetes/app2 has no endpoints
2023-02-02T06:48:37.135702Z	info	ads	Push debounce stable[20] 1 for config ServiceEntry/sample/helloworld.sample.svc.cluster.local: 100.239815ms since last change, 100.23966ms since last push, full=false
2023-02-02T06:48:37.135786Z	info	ads	XDS: Incremental Pushing:2023-02-02T06:48:26Z/9 ConnectedEndpoints:3 Version:2023-02-02T06:48:26Z/9
2023-02-02T06:48:37.830452Z	info	ads	ADS: new connection for node:helloworld-v2-79bf565586-f44h7.sample-4
2023-02-02T06:48:37.831354Z	info	ads	CDS: PUSH request for node:helloworld-v2-79bf565586-f44h7.sample resources:38 size:38.1kB cached:0/33
2023-02-02T06:48:37.846193Z	info	ads	EDS: PUSH request for node:helloworld-v2-79bf565586-f44h7.sample resources:32 size:8.7kB empty:0 cached:32/32
2023-02-02T06:48:37.859465Z	info	ads	LDS: PUSH request for node:helloworld-v2-79bf565586-f44h7.sample resources:36 size:99.0kB
2023-02-02T06:48:37.911556Z	info	ads	RDS: PUSH request for node:helloworld-v2-79bf565586-f44h7.sample resources:18 size:10.1kB cached:0/18
2023-02-02T06:48:37.911578Z	warn	**ads	ADS:LDS: ACK ERROR helloworld-v2-79bf565586-f44h7.sample-4 Internal:Error adding/updating listener(s) 192.168.2.127_9100: error adding listener: '192.168.2.127_9100' has duplicate address '192.168.2.127:9100,[2001:2:127:1::ab06]:9100' as existing listener**

### Version

```prose
istioctl version:
client version: 1.16.0
control plane version: 1.16.2
kubernetes version: v1.24.0
```

**Comment by hzxuzhonghu** [comment]
@zhlsunshine

**Comment by zhlsunshine** [comment]
Hi @hzxuzhonghu checking.

@soft-way how about to apply this [PR#43099](https://github.com/istio/istio/pull/43099) in your case? because I just found a bug for the dual stack support.

**Comment by zhlsunshine** [comment]
BTW @soft-way, what do you get if executing commond `kubectl get endpoints -n prometheus-node-exporter -oyaml`

**Comment by soft-way** [comment]
@zhlsunshine Unfortunatly, it does not fix my issue.  Below is the silimar output for get endpoints:
apiVersion: v1
items:
- apiVersion: v1
  kind: Endpoints
  metadata:
    annotations:
      endpoints.kubernetes.io/last-change-trigger-time: "2023-02-06T09:58:45Z"
    creationTimestamp: "2023-02-06T09:58:45Z"
    labels:
      kubernetes.io/name: NodeExporter
      service.kubernetes.io/headless: ""
    name: node-exporter
    namespace: prometheus-node-exporter
    resourceVersion: "11458"
    uid: 291e5461-a81d-48a3-ba00-a6939af74c24
  subsets:
  - addresses:
    - ip: 192.168.2.127
      nodeName: app2
      targetRef:
        kind: Pod
        name: node-exporter-bd2x8
        namespace: prometheus-node-exporter
        uid: 27458227-5ada-418b-96c9-3dc6be936c55
    ports:
    - name: metrics
      port: 9100
      protocol: TCP
kind: List
metadata:
  resourceVersion: ""

BTW:
If the clusterIP of service node-exporter is not set as "None", it's OK. no duplicate address error. 

I tried another scenario to produce similar issue. This time the conflict IPv6 address is from another service.
1.  Istiod reported error:
2023-02-06T09:58:45.482571Z     debug   Build 19 listeners for node helloworld-v2-79bf565586-62b4d.sample including 17 outbound, 0 http proxy, 1 virtual outbound
2023-02-06T09:58:45.482988Z     info    ads     LDS: PUSH for node:helloworld-v2-79bf565586-62b4d.sample resources:19 size:59.9kB
2023-02-06T09:58:45.483270Z     info    ads     RDS: PUSH for node:helloworld-v2-79bf565586-62b4d.sample resources:8 size:4.8kB cached:7/8
2023-02-06T09:58:45.488183Z     info    ads     EDS: PUSH request for node:istio-ingressgateway-fcc6d77b7-hkgfx.istio-system resources:1 size:77B empty:0 cached:0/1 filtered:16
2023-02-06T09:58:45.495678Z     info    ads     EDS: PUSH request for node:istio-eastwestgateway-6d4b46d958-ftflk.istio-system resources:2 size:157B empty:0 cached:0/2 filtered:32
2023-02-06T09:58:45.508976Z     warn    ads     ADS:LDS: ACK ERROR helloworld-v2-79bf565586-62b4d.sample-18 Internal:Error adding/updating listener(s) 2022:192:168::156_9100: error adding listener: '2022:192:168::156_9100' has duplicate address '[2022:192:168::156]:9100,[**2001:0:156:1::1cdc**]:9100' as existing listener

2023-02-06T09:58:45.514528Z     info    ads     EDS: PUSH request for node:istio-eastwestgateway-85b6d4f75b-g2h9k.istio-system resources:2 size:386B empty:0 cached:0/2 filtered:32

2. Above IPv6 address **2001:0:156:1::1cdc** is clusterIP of another service. service info as below:
apiVersion: v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"name":"nginx","namespace":"test"},"spec":{"ipFamilies":["IPv4","IPv6"],"ipFamilyPolicy":"PreferDualStack","ports":[{"name":"http","port":80,"protocol":"TCP","targetPort":80}],"selector":{"app":"nginx"},"type":"ClusterIP"}}
  creationTimestamp: "2023-02-06T09:02:45Z"
  name: nginx
  namespace: test
  resourceVersion: "6141"
  uid: 7130676f-f6a6-4894-a680-a779d380ae3b
spec:
  clusterIP: 10.10.69.224
  clusterIPs:
  - 10.10.69.224
  - **2001:0:156:1::1cdc**
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: nginx
  sessionAffinity: None
  type: ClusterIP

**Comment by zhlsunshine** [comment]
Hi @soft-way, thanks for your information, it the seems situation in my environment showing as below:
![image](https://user-images.githubusercontent.com/4101246/217118047-bb0c2bc2-1bcc-46a7-9578-ca57694d222e.png)

So I think we should figure out firstly that:
**The service and pods have both IPv4 and IPv6 addresses, but why are there only IPv4 addresses for k8s endpoint?**

**Comment by zhlsunshine** [comment]
Hi @soft-way, I noticed that prometheus node-exporter uses `ListenAndServe` in `github.com/prometheus/exporter-toolkit/web` to start the HTTP server, and its definition as below:
```
// ListenAndServe starts the server on addresses given in WebListenAddresses in
// the FlagConfig or instead uses systemd socket activated listeners if
// WebSystemdSocket in the FlagConfig is true. The FlagConfig is also passed on
// to ServeMultiple.
```
And in node_exporter, it adds `:9100` as the `WebListenAddresses`. So there should be only one socket(ipv4 or ipv6), and I think this can help to explain why there is only one endpoint for node_exporter.
So I have an assumption for this: only 1 endpoint for 2 ip addresses per pod, that's may be the reason for such duplicated addresses in existing listener in only 1 listen port.

So my basic idea is that `node_exporter` can support separately IPv4 and IPv6, but not dual stack. Similarly, the `helloworld` sample as well.

**Comment by zhlsunshine** [comment]
`If the clusterIP of service node-exporter is not set as "None", it's OK. no duplicate address error.`
And this need more investigation.

**Comment by zhlsunshine** [comment]
Hi @soft-way  I deployed the `node_exporter` and `helloworld` in my cluster, it's all okay showing as below:
![image](https://user-images.githubusercontent.com/4101246/217141119-267723c4-004b-4eb5-90c3-437aad8770da.png)

node_exporter service yaml:
![image](https://user-images.githubusercontent.com/4101246/217141275-1417664d-3cc2-4dec-843a-a56ca1a093bf.png)

helloworld service yaml:
![image](https://user-images.githubusercontent.com/4101246/217141198-67e101d8-185d-4500-ac69-192065525267.png)

So I have another question: how did you install the Istio with dual stack enabled?

**Comment by soft-way** [comment]
@zhlsunshine 
Why do you need to make the Cluster IP of node_exporter service being set as None?
>>> Host network was used in Pod of node_exporter service. So clusterIP is not required. Service endpoint will use node IP directly.

Reproduce steps:
1. Following steps in https://istio.io/latest/docs/setup/install/multicluster/multi-primary_multi-network/
Setup istio primary cluster on one node k8s on VM155(CTX_CLUSTER1), remote cluser on another one node k8s on VM156(CTX_CLUSTER2)
run below commands:
kubectl --context="${CTX_CLUSTER1}" create namespace istio-system
kubectl --context="${CTX_CLUSTER1}" label namespace istio-system topology.istio.io/network=network1

# create primary cluster
istioctl install --set values.pilot.env.EXTERNAL_ISTIOD=true --set meshConfig.outboundTrafficPolicy.mode=REGISTRY_ONLY \
    --context="${CTX_CLUSTER1}" -f cluster1.yaml -y

# easywest gateway
cat eastwest1.yaml | istioctl --context="${CTX_CLUSTER1}" install -y -f -
# expose cluster1 control plane
kubectl apply --context="${CTX_CLUSTER1}" -n istio-system -f \
    samples/multicluster/expose-istiod.yaml

# change service istio-eastwestgateway to PreferDualStack
...
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 10.10.165.171
  clusterIPs:
  - 10.10.165.171
  - 2001:0:155:2::34d4
  externalIPs:
  - 192.168.0.155
  - 2022:192:168::155
  externalTrafficPolicy: Cluster
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  **ipFamilyPolicy: PreferDualStack**
  ports:
  - name: status-port
    nodePort: 30164
    port: 15021
    protocol: TCP
    targetPort: 15021
...

# expose cluster1 service
kubectl --context="${CTX_CLUSTER1}" apply -n istio-system -f \
    samples/multicluster/expose-services.yaml

# remote cluster
kubectl --context="${CTX_CLUSTER2}" create namespace istio-system
kubectl --context="${CTX_CLUSTER2}" annotate namespace istio-system topology.istio.io/controlPlaneClusters=cluster1
kubectl --context="${CTX_CLUSTER2}" label namespace istio-system topology.istio.io/network=network2
istioctl install --context="${CTX_CLUSTER2}" -f cluster2.yaml -y

# create eastwest gateway on remote cluster
 cat eastwest2.yaml | istioctl --context="${CTX_CLUSTER2}" install -y -f - 
kubectl --context="${CTX_CLUSTER2}" apply -n istio-system -f \
    samples/multicluster/expose-services.yaml

istioctl x create-remote-secret --context="${CTX_CLUSTER2}" --name=cluster2 | kubectl apply -f - --context="${CTX_CLUSTER1}"

2. install some service on remote cluster
kubectl apply -f prometheus-node-exporter_node-exporter-ds.yaml
kubectl apply -f prometheus-node-exporter_node-exporter-svc.yaml
kubectl create ns nginx
kubectl apply -n nginx -f nginx-dual-stack-svc.yaml


3. install test helloworld v2 on remote cluster
kubectl create --context="${CTX_CLUSTER2}" namespace sample
kubectl label --context="${CTX_CLUSTER2}" namespace sample \
    istio-injection=enabled
kubectl apply --context="${CTX_CLUSTER2}" \
    -f samples/helloworld/helloworld.yaml \
    -l service=helloworld -n sample

kubectl apply --context="${CTX_CLUSTER2}" \
    -f samples/helloworld/helloworld.yaml \
    -l version=v2 -n sample


then found below error in istiod:
2023-02-07T11:09:38.665829Z     debug   attached 1 network filters to listener "192.168.0.156_9100" filter chain 0
2023-02-07T11:09:38.665872Z     debug   attached HTTP filter with 5 http_filter options to listener "192.168.0.156_9100" filter chain 1
2023-02-07T11:09:38.665884Z     debug   buildSidecarOutboundListeners: multiple filter chain listener 192.168.0.156_9100 with 2 chains
2023-02-07T11:09:38.665908Z     debug   attached 1 network filters to listener "2022:192:168::156_9100" filter chain 0
2023-02-07T11:09:38.665951Z     debug   attached HTTP filter with 5 http_filter options to listener "2022:192:168::156_9100" filter chain 1
2023-02-07T11:09:38.665963Z     debug   buildSidecarOutboundListeners: multiple filter chain listener 2022:192:168::156_9100 with 2 chains
2023-02-07T11:09:38.666982Z     debug   Build 20 listeners for node helloworld-v2-79bf565586-w8mqg.sample including 18 outbound, 0 http proxy, 1 virtual outbound
2023-02-07T11:09:38.667325Z     info    ads     LDS: PUSH request for node:helloworld-v2-79bf565586-w8mqg.sample resources:20 size:63.0kB
2023-02-07T11:09:38.749359Z     info    ads     RDS: PUSH request for node:helloworld-v2-79bf565586-w8mqg.sample resources:9 size:6.1kB cached:7/9
2023-02-07T11:09:38.749398Z     warn    ads     ADS:LDS: ACK ERROR helloworld-v2-79bf565586-w8mqg.sample-14 Internal:Error adding/updating listener(s) 192.168.0.156_9100: error adding listener: '192.168.0.156_9100' has duplicate address '192.168.0.156:9100,[2001:0:156:2::dbf]:9100' as existing listener

kubectl get svc -n nginx nginx -oyaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"name":"nginx","namespace":"nginx"},"spec":{"ipFamilies":["IPv4","IPv6"],"ipFamilyPolicy":"PreferDualStack","ports":[{"name":"http","port":80,"protocol":"TCP","targetPort":80}],"selector":{"app":"nginx"},"type":"ClusterIP"}}
  creationTimestamp: "2023-02-07T10:34:57Z"
  name: nginx
  namespace: nginx
  resourceVersion: "3792"
  uid: 374b48c7-d7a4-4ab8-ab1a-be72446898a6
spec:
  clusterIP: 10.10.212.79
  clusterIPs:
  - 10.10.212.79
  - **2001:0:156:2::dbf**
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: nginx
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}


kubectl get svc -n prometheus-node-exporter   node-exporter -oyaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{"prometheus.io/scrape":"true"},"labels":{"kubernetes.io/name":"NodeExporter"},"name":"node-exporter","namespace":"prometheus-node-exporter"},"spec":{"clusterIP":"None","clusterIPs":["None"],"internalTrafficPolicy":"Cluster","ipFamilies":["IPv4","IPv6"],"ipFamilyPolicy":"PreferDualStack","ports":[{"name":"metrics","port":9100,"protocol":"TCP","targetPort":9100}],"selector":{"app":"prometheus","module":"node-exporter"},"sessionAffinity":"None","type":"ClusterIP"}}
    prometheus.io/scrape: "true"
  creationTimestamp: "2023-02-07T10:35:14Z"
  labels:
    kubernetes.io/name: NodeExporter
  name: node-exporter
  namespace: prometheus-node-exporter
  resourceVersion: "3859"
  uid: f1f25ad8-512c-483d-b073-4af0dece6cc2
spec:
  clusterIP: None
  clusterIPs:
  - None
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - name: metrics
    port: 9100
    protocol: TCP
    targetPort: 9100
  selector:
    app: prometheus
    module: node-exporter
  sessionAffinity: None
  type: ClusterIP
status:
  loadBalancer: {}

eastwest gateway service on primary cluster:
kubectl get svc -n istio-system  istio-eastwestgateway -oyaml
apiVersion: v1
kind: Service
metadata:
  annotations:
    kubectl.kubernetes.io/last-applied-configuration: |
      {"apiVersion":"v1","kind":"Service","metadata":{"annotations":{},"labels":{"app":"istio-eastwestgateway","install.operator.istio.io/owning-resource":"eastwest","install.operator.istio.io/owning-resource-namespace":"istio-system","istio":"eastwestgateway","istio.io/rev":"default","operator.istio.io/component":"IngressGateways","operator.istio.io/managed":"Reconcile","operator.istio.io/version":"1.16.2","release":"istio","topology.istio.io/network":"network1"},"name":"istio-eastwestgateway","namespace":"istio-system"},"spec":{"externalIPs":["192.168.0.155","2022:192:168::155"],"ports":[{"name":"status-port","port":15021,"protocol":"TCP","targetPort":15021},{"name":"tls","port":15443,"protocol":"TCP","targetPort":15443},{"name":"tls-istiod","port":15012,"protocol":"TCP","targetPort":15012},{"name":"tls-webhook","port":15017,"protocol":"TCP","targetPort":15017}],"selector":{"app":"istio-eastwestgateway","istio":"eastwestgateway","topology.istio.io/network":"network1"},"type":"LoadBalancer"}}
  creationTimestamp: "2023-02-07T10:08:36Z"
  labels:
    app: istio-eastwestgateway
    install.operator.istio.io/owning-resource: eastwest
    install.operator.istio.io/owning-resource-namespace: istio-system
    istio: eastwestgateway
    istio.io/rev: default
    operator.istio.io/component: IngressGateways
    operator.istio.io/managed: Reconcile
    operator.istio.io/version: 1.16.2
    release: istio
    topology.istio.io/network: network1
  name: istio-eastwestgateway
  namespace: istio-system
  resourceVersion: "5127"
  uid: dcc70e02-b24b-43bd-9e7d-10f351305b51
spec:
  allocateLoadBalancerNodePorts: true
  clusterIP: 10.10.165.171
  clusterIPs:
  - 10.10.165.171
  - 2001:0:155:2::34d4
  externalIPs:
  - 192.168.0.155
  - 2022:192:168::155
  externalTrafficPolicy: Cluster
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - name: status-port
    nodePort: 30164
    port: 15021
    protocol: TCP
    targetPort: 15021
  - name: tls
    nodePort: 31641
    port: 15443
    protocol: TCP
    targetPort: 15443
  - name: tls-istiod
    nodePort: 30988
    port: 15012
    protocol: TCP
    targetPort: 15012
  - name: tls-webhook
    nodePort: 31085
    port: 15017
    protocol: TCP
    targetPort: 15017
  selector:
    app: istio-eastwestgateway
    istio: eastwestgateway
    topology.istio.io/network: network1
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer: {}

**Comment by soft-way** [comment]
yaml file for above steps:
cluster1.yaml:
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  values:
    global:
      meshID: mesh1
      multiCluster:
        clusterName: cluster1
      network: network1


eastwest1.yaml:
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: eastwest
spec:
  revision: ""
  profile: empty
  components:
    ingressGateways:
      - name: istio-eastwestgateway
        label:
          istio: eastwestgateway
          app: istio-eastwestgateway
          topology.istio.io/network: network1
        enabled: true
        k8s:
          env:
            # traffic through this gateway should be routed inside the network
            - name: ISTIO_META_REQUESTED_NETWORK_VIEW
              value: network1
          service:
            ports:
              - name: status-port
                port: 15021
                targetPort: 15021
              - name: tls
                port: 15443
                targetPort: 15443
              - name: tls-istiod
                port: 15012
                targetPort: 15012
              - name: tls-webhook
                port: 15017
                targetPort: 15017
            externalIPs:
              - 192.168.0.155
              - 2022:192:168::155
  values:
    gateways:
      istio-ingressgateway:
        injectionTemplate: gateway
    global:
      network: network1


cluster2.yaml:
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
spec:
  profile: remote
  values:
    istiodRemote:
      injectionPath: /inject/cluster/cluster2/net/network2
    global:
      remotePilotAddress: 192.168.0.155


 eastwest2.yaml:
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  name: eastwest
spec:
  revision: ""
  profile: empty
  components:
    ingressGateways:
      - name: istio-eastwestgateway
        label:
          istio: eastwestgateway
          app: istio-eastwestgateway
          topology.istio.io/network: network2
        enabled: true
        k8s:
          env:
            # traffic through this gateway should be routed inside the network
            - name: ISTIO_META_REQUESTED_NETWORK_VIEW
              value: network2
          service:
            ports:
              - name: status-port
                port: 15021
                targetPort: 15021
              - name: tls
                port: 15443
                targetPort: 15443
              - name: tls-istiod
                port: 15012
                targetPort: 15012
              - name: tls-webhook
                port: 15017
                targetPort: 15017
            externalIPs:
              - 192.168.0.156
              - 2022:192:168::156
  values:
    gateways:
      istio-ingressgateway:
        injectionTemplate: gateway
    global:
      network: network2

prometheus-node-exporter_node-exporter-ds.yaml:
apiVersion: apps/v1
kind: DaemonSet
metadata:
  annotations:
  labels:
    app: prometheus
    module: node-exporter
  name: node-exporter
  namespace: prometheus-node-exporter
spec:
  selector:
    matchLabels:
      app: prometheus
      module: node-exporter
  template:
    metadata:
      labels:
        app: prometheus
        module: node-exporter
    spec:
      containers:
      - args:
        - --path.procfs=/host/proc
        - --path.sysfs=/host/sys
        image: prom/node-exporter:v1.3.1
        imagePullPolicy: IfNotPresent
        name: prometheus-node-exporter
        ports:
        - containerPort: 9100
          hostPort: 9100
          name: metrics
          protocol: TCP
        resources:
          limits:
            cpu: "1"
            memory: 1Gi
          requests:
            cpu: 10m
            memory: 50Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /host/proc
          name: proc
          readOnly: true
        - mountPath: /host/sys
          name: sys
          readOnly: true
      dnsPolicy: ClusterFirst
      hostNetwork: true
      hostPID: true
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext: {}
      terminationGracePeriodSeconds: 30
      volumes:
      - hostPath:
          path: /proc
          type: ""
        name: proc
      - hostPath:
          path: /sys
          type: ""
        name: sys
  updateStrategy:
    rollingUpdate:
      maxSurge: 0
      maxUnavailable: 1
    type: RollingUpdate
	
prometheus-node-exporter_node-exporter-svc.yaml:
apiVersion: v1
kind: Service
metadata:
  annotations:
    prometheus.io/scrape: "true"
  labels:
    kubernetes.io/name: NodeExporter
  name: node-exporter
  namespace: prometheus-node-exporter
spec:
  clusterIP: None
  clusterIPs:
  - None
  internalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - name: metrics
    port: 9100
    protocol: TCP
    targetPort: 9100
  selector:
    app: prometheus
    module: node-exporter
  sessionAffinity: None
  type: ClusterIP

nginx-dual-stack-svc.yaml:
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1.23.2
        ports:
        - name: http
          containerPort: 80

---
apiVersion: v1
kind: Service
metadata:
  name: nginx
spec:
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: nginx
  type: ClusterIP

**Comment by zhlsunshine** [comment]
Hi @soft-way, I will try my best to reproduce your case. However, it's unreasonable to set `ipFamilyPolicy` as `PreferDualStack` and `ClusterIP` as `None`. 
If I try to deploy such service, I got below message:
![image](https://user-images.githubusercontent.com/4101246/217408021-06abf152-7563-48be-9d09-f5c77a34cdb3.png)

It means that it's abnormal for kubernetes to deploy such service.

**Comment by zhlsunshine** [comment]
Hi @soft-way, I will try my best to reproduce your case. However, for your answer:

```
Host network was used in Pod of node_exporter service. So clusterIP is not required. Service endpoint will use node IP directly.
```
As we have tried, the `node_exporter` and `helloworld` service can support separately IPv4 and IPv6, but can not support dual stack(IPv4 and IPv6 at the same time), so only one endpoint of them can be there. Beside that, there are some enhancement for headless service support on dual stack. Please track the PR: https://github.com/istio/istio/pull/41248 for more information.

**Comment by soft-way** [comment]
Hi @zhlsunshine , I tried the endpoint node-exporter, The IPv6 can be added by "kubectl edit" command.
![k8s-endpoints-add-ipv6](https://user-images.githubusercontent.com/12678582/217447207-a25a1d6a-bddd-4cf9-93de-6ee1247ad1b4.png)

It seems that k8s will not auto add IPv6 address into endpoints for dualstack service:
![k8s-endpoints-ipv6](https://user-images.githubusercontent.com/12678582/217447407-5855ba0b-7d1f-477a-a0fe-7d887f6d9166.png)

BTW, I uploaded yaml files and dump log files used in my test in above comment.

**Comment by zhlsunshine** [comment]
Hi @soft-way, I think you may need to check `endpointslice` via `kubectl get endpointslice` for kubernetes. And for headless service, I have mentioned that. Let's go on tracking it.

**Comment by soft-way** [comment]
Hi @zhlsunshine kubectl get endpointslice  -n prometheus-node-exporter:
![kubectl-get-endpointslice](https://user-images.githubusercontent.com/12678582/217451372-c63de708-40d0-4335-8aaf-c2ad110e39a1.png)

**Comment by soft-way** [comment]
From istio-proxy side car log (file: helloworld-v2_istio-proxy.out).
![duplicate-address](https://user-images.githubusercontent.com/12678582/217451748-547ac211-7a1a-4854-844f-890bcecdd607.png)

**Comment by zhlsunshine** [comment]
Hi @soft-way, yeah, that fix (I mean [PR#41248](https://github.com/istio/istio/pull/41248)) is flawed, so I reverted it, and need some other changes.
